// @juneil/entityts/lib/enums.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var ModeEnum;
(function (ModeEnum) {
    ModeEnum["READ"] = "mode_read";
    ModeEnum["CREATE"] = "mode_create";
    ModeEnum["UPDATE"] = "mode_update";
})(ModeEnum = exports.ModeEnum || (exports.ModeEnum = {}));
var TypeEnum;
(function (TypeEnum) {
    TypeEnum[TypeEnum["Any"] = Symbol('type_any')] = "Any";
    TypeEnum[TypeEnum["ObjectId"] = Symbol('type_objectid')] = "ObjectId";
    TypeEnum[TypeEnum["Hex"] = Symbol('type_hex')] = "Hex";
    TypeEnum[TypeEnum["Base64"] = Symbol('type_base64')] = "Base64";
    TypeEnum[TypeEnum["IsoDate"] = Symbol('type_isodate')] = "IsoDate";
    TypeEnum[TypeEnum["Integer"] = Symbol('type_integer')] = "Integer";
    TypeEnum[TypeEnum["IP"] = Symbol('type_ip')] = "IP";
    TypeEnum[TypeEnum["URI"] = Symbol('type_uri')] = "URI";
    TypeEnum[TypeEnum["Email"] = Symbol('type_email')] = "Email";
    TypeEnum[TypeEnum["Timestamp"] = Symbol('type_timestamp')] = "Timestamp";
})(TypeEnum = exports.TypeEnum || (exports.TypeEnum = {}));
